### PR TITLE
kv,store/copr: make kv.KeyRange the same with coprocessor.KeyRange, so we can use unsafe optimization

### LINKED ISSUE #37286: Reduce coprocessor request cost
## Enhancement

image

https://github.com/pingcap/tidb/blob/21847fe58b51807696debf0f7650af948f2aa3e2/store/copr/key_ranges.go#L137-L144

Since kv.KeyRange and coprocessor.KeyRange definition are different, we need to make object allocation to send coprocessor.Request...

We can avoid object allocation if `kv.KeyRange` and `coprocessor.KeyRange` are the same.
Actually, they are both 

```
type KeyRange struct {
    StartKey []byte
    EndKey []byte
}
```

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve?

Issue Number: close #37286

Problem Summary:

### What is changed and how it works?

make `kv.KeyRange` the same with `coprocessor.KeyRange` so we avoid converting and thus reduce object allocation.

The same workload, get the pprof

Before:

After:

image

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [X] Manual test (add detailed scripts or steps below)

TPCH 50
Execute Query 16

```
explain analyze select p_brand, p_type, p_size, count(distinct ps_suppkey) as supplier_cnt from partsupp, part where p_partkey = ps_partkey and p_brand <> 'Brand#34' and p_type not like 'LARGE BRUSHED%' and p_size in (48, 19, 12, 4, 41, 7, 21, 39) and ps_suppkey not
in ( select s_suppkey from supplier where s_comment like '%Customer%Complaints%' ) group by p_brand, p_type, p_size order by supplier_cnt desc, p_brand, p_type, p_size;
```

Get the profile flamegraph.

- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by disksing — APPROVED** [review]

File: kv/key.go [hunk 1]
```diff
@@ -89,9 +91,26 @@ func (k Key) String() string {
...
+	XXX_sizecache        int32
+}
+
+func init() {
+	// kv.KeyRange and coprocessor.KeyRange should be the same layout, so we can exchange them.
+	var r1 KeyRange
+	var r2 coprocessor.KeyRange
+	if unsafe.Sizeof(r1) != unsafe.Sizeof(r2) {
+		panic("coprocessor protocol change?")
+	}
```
Comment (xhebox) [inline kv/key.go:113]: I think we can make it a test. Also  use `require.Equal` to test if their default value is exactly same will be better.

Comment (disksing) [reply kv/key.go:113]: good point.

Comment (tiancaiamao) [reply kv/key.go]: Done

**Review by xhebox — APPROVED** [review]

**Review by disksing — APPROVED** [review]

**Review by hawkingrei — APPROVED** [review]

**Pull request merged**